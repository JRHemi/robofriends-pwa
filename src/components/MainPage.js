import React, { Component } from 'react';

import CardList from './CardList';
import SearchBox from './SearchBox';
import Scroll from './Scroll';
import Header from "./Header"
import ErrorBoundary from './ErrorBoundary';

import './MainPage.css';

class MainPage extends Component {
  componentDidMount() {
    this.props.onRequestRobots();
  }

  filterRobots = () => {
    const filteredRobots = this.props.robots.filter(robot => {
        return robot.name.toLowerCase().includes(this.props.searchField.toLowerCase());
      })

    return filteredRobots
  }

  render() {
    const { robots, onSearchChange, isPending } = this.props;

    return (
      <div className='tc'>
        <Header />
        <SearchBox searchChange={onSearchChange}/>
        <Scroll>
          { isPending ? <h1>Loading</h1> :
            <ErrorBoundary>
              <CardList robots={this.filterRobots()} />
            </ErrorBoundary>
          }
        </Scroll>
      </div>
    );
  }
}

// action done from mapDispatchToProps will change state from mapStateToProps
export default MainPage
