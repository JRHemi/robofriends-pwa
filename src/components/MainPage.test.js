import React from "react";
import { shallow } from "enzyme";
import MainPage from "./MainPage";

let wrapper;
beforeEach(() => {
  const mockProps = {
    onRequestRobots: jest.fn(),
    robots: [],
    searchField: "",
    isPending: false,
  };

  wrapper = shallow(<MainPage {...mockProps} />);
});

describe("MainPage Component", () => {
  // Ensure mocks are cleared after each test
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("expects to render MainPage component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("displays loading screen", () => {
    const mockPropsPendingResponse = {
      onRequestRobots: jest.fn(),
      robots: [],
      searchField: "",
      isPending: true,
    };

    const wrapperPending = shallow(<MainPage {...mockPropsPendingResponse} />);
    expect(wrapperPending).toMatchSnapshot();
  });

  it("filters robots correct search", () => {
    const mockPropsCorrectSearch = {
      onRequestRobots: jest.fn(),
      robots: [
        {
          id: 3,
          name: "john",
          email: "john@gamil.com",
        },
      ],
      searchField: "john",
      isPending: false,
    };

    const filteredRobotsCorrectSearch = mockPropsCorrectSearch.robots;
    const wrapper2 = shallow(<MainPage {...mockPropsCorrectSearch} />);
    expect(wrapper2.instance().filterRobots()).toEqual(
      filteredRobotsCorrectSearch
    );
  });

  it("filters robots incorrect search", () => {
    const mockPropsPopulatedIncorrectSearch = {
      onRequestRobots: jest.fn(),
      robots: [
        {
          id: 3,
          name: "john",
          email: "john@gamil.com",
        },
      ],
      searchField: "aaa",
      isPending: false,
    };

    const filteredRobotsIncorrectSearch = [];
    const wrapper3 = shallow(
      <MainPage {...mockPropsPopulatedIncorrectSearch} />
    );
    expect(wrapper.instance().filterRobots()).toEqual([]);
    expect(wrapper3.instance().filterRobots()).toEqual(
      filteredRobotsIncorrectSearch
    );
  });
});
