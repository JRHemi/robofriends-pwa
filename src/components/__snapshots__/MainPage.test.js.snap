// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage Component displays loading screen 1`] = `
<div
  className="tc"
>
  <Header />
  <SearchBox />
  <Scroll>
    <h1>
      Loading
    </h1>
  </Scroll>
</div>
`;

exports[`MainPage Component expects to render MainPage component 1`] = `
<div
  className="tc"
>
  <Header />
  <SearchBox />
  <Scroll>
    <ErrorBoundary>
      <CardList
        robots={Array []}
      />
    </ErrorBoundary>
  </Scroll>
</div>
`;
